mesonbuild/interpreter/interpreter.py

```diff
@@ -3039,39 +3039,34 @@ def add_target(self, name: str, tobj: build.Target) -> None:
 
     @FeatureNew('both_libraries', '0.46.0')
     def build_both_libraries(self, node, args, kwargs):
-        shared_lib = self.build_target(node, args, kwargs, build.SharedLibrary)
         static_lib = self.build_target(node, args, kwargs, build.StaticLibrary)
 
-        # Check if user forces non-PIC static library.
-        pic = True
-        key = OptionKey('b_staticpic')
-        if 'pic' in kwargs:
-            pic = kwargs['pic']
-        elif key in self.environment.coredata.options:
-            pic = self.environment.coredata.options[key].value
-
         if self.backend.name == 'xcode':
             # Xcode is a bit special in that you can't (at least for the moment)
             # form a library only from object file inputs. The simple but inefficient
             # solution is to use the sources directly. This will lead to them being
             # built twice. This is unfortunate and slow, but at least it works.
             # Feel free to submit patches to get this fixed if it is an
             # issue for you.
+            #
+            # XXX: since we switched from including object files to using link_whole,
+            # This comment (and code), maybe out of date. I don't have a machine with
+            # xcode to test with, however.
             reuse_object_files = False
         else:
-            reuse_object_files = pic
+            reuse_object_files = static_lib.pic
 
         if reuse_object_files:
-            # Replace sources with objects from the shared library to avoid
-            # building them twice. We post-process the static library instead of
-            # removing sources from args because sources could also come from
-            # any InternalDependency, see BuildTarget.add_deps().
-            static_lib.objects.append(build.ExtractedObjects(shared_lib, shared_lib.sources, shared_lib.generated, []))
-            static_lib.sources = []
-            static_lib.generated = []
-            # Compilers with no corresponding sources confuses the backend.
-            # Keep only compilers used for linking
-            static_lib.compilers = {k: v for k, v in static_lib.compilers.items() if k in compilers.clink_langs}
+            # If we can reuse the object files from the static library, then we
+            # can simply `link_whole` with it, and use it's include_directories.
+            # We use the include_directoreis from the static_lib since it's
+            # already unrolled any include directories added by a dependency.
+            shared_lib = self.build_target(
+                node, [args[0]],
+                {'link_whole': [static_lib], 'include_directories': static_lib.get_include_dirs()},
+                build.SharedLibrary)
```

Comment:
You need to pass most other kwargs too e.g. `install`, etc. `link_args` is tricky, because those could come from dependencies.

Why do you need include directories? This target should not be building any sources, so that should be useless, no?